### PR TITLE
[fix] Catch closed upgrade requests in `uws`

### LINKED ISSUE #530: uws: Cannot read property 'fd' of null
**Version:** 6.0.7

**Environment:**
  - **Operating system:** Linux
  - **Node.js:** 6.9.4

Running Linux; NodeJS via Primus, no SSL.

Stack:

```
/srv/node/node_modules/bluebird/js/release/async.js:61
        fn = function () { throw arg; };
                           ^
TypeError: Cannot read property 'fd' of null
    at child.on (/srv/node/node_modules/primus/transformers/uws/server.js:90:21)
    at child.emit (/srv/node/node_modules/primus/node_modules/eventemitter3/index.js:131:35)
    at emitter (/srv/node/node_modules/emits/index.js:26:22)
    at iterate (/srv/node/node_modules/primus/transformer.js:125:24)
    at done (/srv/node/node_modules/primus/transformer.js:139:7)
    at authorized (/srv/node/node_modules/primus/middleware/authorization.js:22:24)
    at tryCatcher (/srv/node/node_modules/bluebird/js/release/util.js:16:23)
    at Promise.successAdapter (/srv/node/node_modules/bluebird/js/release/nodeify.js:22:30)
    at Promise.<anonymous> (/srv/node/node_modules/continuation-local-storage/context.js:84:17)
    at bound (domain.js:280:14)
    at Promise.runBound [as _fulfillmentHandler0] (domain.js:293:12)
    at Promise._settlePromise (/srv/node/node_modules/bluebird/js/release/promise.js:564:21)
    at Promise._settlePromise0 (/srv/node/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/srv/node/node_modules/bluebird/js/release/promise.js:691:18)
    at Async._drainQueue (/srv/node/node_modules/bluebird/js/release/async.js:138:16)
    at Async._drainQueues (/srv/node/node_modules/bluebird/js/release/async.js:148:10)
    at Immediate.Async.drainQueues (/srv/node/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:637:20)
    at tryOnImmediate (timers.js:610:5)
    at processImmediate [as _immediateCallback] (timers.js:582:5)
```

This seems to happen at random on a production server; seems to be something like a connection being disconnected before there's a chance to register it - not sure.

The [offending code](https://github.com/primus/primus/blob/7209e8ac472ec23980b8c53ccd8d31c20a243911/transformers/uws/server.js#L90) (appears `socketHandle = soc._handle` is null, not using SSL):

```js
      const ticket = native.transfer(
        socketHandle.fd === -1 ? socketHandle : socketHandle.fd,
        sslState
      );
```

Unfortunately I don't have reproduction steps; we have as of yet been unable to trace down the cause. Any ideas?

### PR DESCRIPTION
**Pull request by STRML**
Fixes #530 

Error handler and checking of readable/writable state is very similar to how this is [done in ws](https://github.com/websockets/ws/blob/95d8553ca657ce926934cd5d8b5eb3c7841f4422/lib/WebSocketServer.js#L150).

### CODE REVIEW

File: transformers/uws/server.js [hunk 1]
```diff
@@ -76,6 +76,16 @@ module.exports = function server() {
     const secKey = req.headers['sec-websocket-key'];
 
     if (secKey && secKey.length === 24) {
+      //
+      // Destroy the socket if it has already been sent a FIN packet.
+      //
+      if (!socket.readable || !socket.writable) return socket.destroy();
```
Comment (lpinca) [inline transformers/uws/server.js]: This is probably the only thing that we need. I think adding an error listener is pointless as the socket is destroyed in the same tick.

Comment (lpinca) [inline transformers/uws/server.js]: Tests fail because the variable is called `soc` not `socket`.

Comment (STRML) [inline transformers/uws/server.js]: My apologies, haven't had my coffee this morning; updated

**Comment by STRML** [comment]
Tests passing

File: transformers/uws/server.js [hunk 2]
```diff
@@ -76,6 +76,11 @@ module.exports = function server() {
     const secKey = req.headers['sec-websocket-key'];
 
     if (secKey && secKey.length === 24) {
```
Comment (lpinca) [inline transformers/uws/server.js]: Can we move the check here?
```js
if (soc.readable && soc.writable && secKey && secKey.length === 24) {
  ...
} 
```

Comment (lpinca) [inline transformers/uws/server.js]: Comment can be removed.

**Comment by STRML** [comment]
:+1:

**Comment by lpinca** [comment]
LGTM will merge it once CI completes.

**Comment by lpinca** [comment]
Thank you!

**Comment by lpinca** [comment]
Released 6.0.8 with this fix.

**Comment by lpinca** [comment]
@suchitagarwal in https://github.com/uWebSockets/uWebSockets/issues/326 you are most likely affected by the issue that has been fixed here.

The socket can be closed before [handleUpgrade](https://github.com/uWebSockets/uWebSockets/blob/dfcc9a9939d2cc512c3a6d3516368f37a8229254/nodejs/dist/uws.js#L446-L470) is invoked and when that happens the [handle](https://github.com/uWebSockets/uWebSockets/blob/dfcc9a9939d2cc512c3a6d3516368f37a8229254/nodejs/dist/uws.js#L455) is `null`.

**Comment by lpinca** [comment]
@silverwind the checks for `.readable` and `.writable` have been ported from `ws `.
The check for `.readable ` is there because the client may have already called `socket.end()` on its side. The check for `.writable` is there because we shouldn't write to a closed socket.

It should be ok to only check if `handle !== null` in uws.

Don't cc me in uWebSockets repo as I'm banned.

**Comment by silverwind** [comment]
Thanks for clarifying, I wasn't aware of that drama 😉.

**Comment by alexhultman** [comment]
Btw @lpinca I removed your ban a while ago.

**Comment by lpinca** [comment]
@alexhultman good to know, thanks.

**Pull request merged**